### PR TITLE
feat: #8227 payments mailer service refactor

### PR DESCRIPTION
**Pull request by willmcvay**
# Pull request checklist

**Detail as per issue below (required):**

fixes: #8227

### CODE REVIEW

File: packages/payments-service/src/core/server.ts [hunk 1]
```diff
@@ -1,34 +1,52 @@
...
+export const bootstrapApplication = async (): Promise<[INestApplication, Express]> => {
+  process.env = {
+    ...process.env,
+    ...config,
+  }
+
+  const expressApp = express()
+  const app = await NestFactory.create(AppModule, new ExpressAdapter(expressApp))
+
+  app.useGlobalPipes(new ValidationPipe())
```
Comment (bashleigh) [inline packages/payments-service/src/core/server.ts:22]: This is where the validationPipe comes from, in future if you want to use `class-validator` you'll need to add this line to your project.

Comment (willmcvay) [reply packages/payments-service/src/core/server.ts:22]: Got you 👍🏻

File: packages/payments-service/src/payment-receipt/provider.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+      paymentReason,
+      recipientName,
+      paymentDate: new Date().toDateString(),
+      paymentAmount: `${paymentAmount.toFixed(2)}`,
+    })
+    const title = `Payment Receipt from ${companyName}`
+
+    const mailId = await sendEmail(receipientEmail, title, template, senderEmail)
+
+    return mailId
```
Comment (bashleigh) [inline packages/payments-service/src/payment-receipt/provider.ts:33]:
```suggestion
    return sendEmail(receipientEmail, title, template, senderEmail)
```

no need for declaration if it's not used, Promises can be returned in an async func

File: packages/payments-service/src/payment-request/controller.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+export class PaymentRequestController {
+  constructor(private readonly paymentRequestProvider: PaymentRequestProvider) {}
+
+  @Post('/:paymentId')
+  async createRequest(
+    @Headers() paymentRequestHeaders: PaymentRequestHeaders,
+    @Param() { paymentId }: PaymentRequestParams,
+    @Body() paymentRequest: PaymentRequestDto,
+  ): Promise<void> {
+    this.paymentRequestProvider.create(paymentRequestHeaders, paymentRequest, paymentId)
```
Comment (bashleigh) [inline packages/payments-service/src/payment-request/controller.ts:15]:
```suggestion
    await this.paymentRequestProvider.create(paymentRequestHeaders, paymentRequest, paymentId)
```

Does this need to be awaited? Not returned this time so presuming you want to await it

File: packages/payments-service/src/payment-request/dto.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  @IsString()
+  @IsNotEmpty()
+  recipientName: string
+
+  @IsString()
+  @IsNotEmpty()
+  paymentReason: string
+
+  @IsNumber()
+  paymentAmount: number
```
Comment (bashleigh) [inline packages/payments-service/src/payment-request/dto.ts:17]: add `@IsNotEmpty()` or `paymentAcount?: number` as could be undefined

File: packages/payments-service/src/payment-request/module.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import { Module } from '@nestjs/common'
+import { ConfigModule } from '@nestjs/config'
+import { DynamoDBModule } from '../dynamo'
+import { PaymentRequestController } from './controller'
+import { PaymentRequestProvider } from './provider'
+
+@Module({
+  imports: [DynamoDBModule.forFeature(), ConfigModule],
```
Comment (bashleigh) [inline packages/payments-service/src/payment-request/module.ts:8]: very niiiiceeeeé

File: packages/payments-service/src/payment-request/provider.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+      url: `${process.env.PAYMENTS_APP_URI}/payments/${paymentId}?session=${session}&clientCode=${clientCode}`,
+      recipientName,
+      paymentExpiry: new Date(paymentExpiry).toDateString(),
+      paymentAmount: `${paymentAmount.toFixed(2)}`,
+    })
+    const title = `Payment Request from ${companyName}`
+
+    const mailId = await sendEmail(receipientEmail, title, template, senderEmail)
+
+    return mailId
```
Comment (bashleigh) [inline packages/payments-service/src/payment-request/provider.ts:37]:
```suggestion
    await sendEmail(receipientEmail, title, template, senderEmail)
```

No need for declaration of unsed var, can return promise in async func

File: packages/payments-service/src/payments/provider.ts [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+import { PaymentWithPropertyModel } from '../types/payment'
+
+@Injectable()
+export class PaymentsProvider {
+  async getPayment(paymentsHeaders: PaymentsHeaders, paymentId: string): Promise<PaymentWithPropertyModel> {
+    const accessToken = await reapitConnectSession.connectAccessToken()
+    const clientCode = paymentsHeaders['reapit-customer']
+    const apiVersion = paymentsHeaders['api-version']
+
+    if (!accessToken) throw new Error('No access token returned from Reapit Connect')
```
Comment (bashleigh) [inline packages/payments-service/src/payments/provider.ts:16]: UnauthorizedException?

**Review by bashleigh — APPROVED** [review]
Looks good! Just a few suggestions and snippets that might help

File: packages/payments-service/src/core/ses-client.ts [hunk 8]
```diff
@@ -2,6 +2,13 @@ import AWS from 'aws-sdk'
 
 const ses = new AWS.SES({ apiVersion: '2010-12-01' })
 
+// TODO - these values will be configurable by client see: https://github.com/reapit/foundations/issues/8228
+export const defaultEmailConfig = {
+  senderEmail: 'Reapit Payments <wmcvay@reapit.com>',
```
[the remark this answers is not in the document]
Comment (willmcvay) [reply packages/payments-service/src/core/ses-client.ts:7]: It is for now, see that GH ticket, the sender mailbox will need to be configured by SI later. This is just for testing.

File: packages/payments-service/src/payment-receipt/controller.ts [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+import { PaymentReceiptDto, PaymentReceiptHeaders } from './dto'
+import { PaymentReceiptProvider } from './provider'
+
+@Controller('receipt/private')
+export class PaymentReceiptPrivateController {
+  constructor(private readonly paymentReceiptProvider: PaymentReceiptProvider) {}
+
+  @Post('/:paymentId')
+  async createReceiptPrivate(
+    @Headers() paymentReceiptHeaders: PaymentReceiptHeaders,
```
[the remark this answers is not in the document]
Comment (willmcvay) [reply packages/payments-service/src/payment-receipt/controller.ts:12]: There is a gateway authorizer that validates the access token before the lamba is invoked. Standard OAuth flow, hence private.

File: packages/payments-service/src/payment-receipt/provider.ts [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+@Injectable()
+export class PaymentReceiptProvider {
+  async createReceipt(
+    paymentReceiptHeaders: PaymentReceiptHeaders,
+    paymentReceipt: PaymentReceiptDto,
+  ): Promise<string> {
+    const { receipientEmail, recipientName, paymentReason, paymentAmount, paymentCurrency } = paymentReceipt
+
+    const clientCode = paymentReceiptHeaders['reapit-customer']
+    console.log(clientCode)
```
Comment (bashleigh) [inline packages/payments-service/src/payment-receipt/provider.ts:16]: console log

**Pull request merged**